utils/ensemble_tools.py

```diff
@@ -0,0 +1,166 @@
+""" A module to create input directories and encode
+configfurations for FabNESO
+"""
+
+import os
+import shutil
+import itertools
+import re
+from pathlib import Path
+import lxml.etree as ET
+
+
+def create_dir_tree(
+    sweep_path,
+    n_dirs,
+    destructive,
+    copy_dir,
+    edit_file,
+    parameter_to_scan,
+    scan_range,
+    outdir_prefix,
+):
+    """Create a directory tree in the sweep_path"""
+
+    copy_files = os.path.isdir(copy_dir)
+    if not copy_files:
+        print("No correct copy dir found, just creating the tree for now!")
+
+    # Make the base directory
+    try:
+        make_directory(sweep_path, destructive)
+    except FileExistsError:
+        return
+
+    # Set the initial value of the scanned parameter to the lower limit
+    para_val = scan_range[0]
+
+    for i in range(n_dirs):
+        new_dir = Path(sweep_path) / "SWEEP" / outdir_prefix + i
+        # Make the directory
+        os.makedirs(new_dir)
+        # If we're copying files, do so
+        if copy_files:
+            copy_dir_contents(new_dir, copy_dir)
+        # Now we edit the parameter file for our
+        # template scan if we're doing that
+        if os.path.isfile(new_dir + edit_file) and parameter_to_scan:
+            edit_parameter(new_dir + edit_file, parameter_to_scan, para_val)
```

Comment:
```suggestion
        if (new_dir / edit_file).is_file() and parameter_to_scan:
            edit_parameter(new_dir / edit_file, parameter_to_scan, para_val)
```
`pathlib.Path` instances should be joined with `/` (they don't support string concatenation with `+`). Also more idiomatic to use `Path` methods rather than `os.path` functions

Reply:
Also I would say we might want to raise an exception if `(new_dir / edit_file).is_file()` is false or `parameter_to_scan` is `None` rather than only call `edit_parameter` if the inverse is true as otherwise we are just creating a sweep directory with verbatim copies of the contents of `copy_dir` which seems to be unlikely what the user would want. Also generally better from a readability and robustness perspective to be explicit when testing if value is empty or `None` rather than relying on empty objects being considered false - so for example `parameter_to_scan != ""` or `parameter_to_scan is not None` rather than just `parameter_to_scan`.

Reply:
Changed these to raise exceptions as suggested